xinference/model/llm/vllm/core.py

```diff
@@ -319,16 +323,42 @@ async def async_generate(
 
         async def stream_results() -> AsyncGenerator[CompletionChunk, None]:
             previous_texts = [""] * sanitized_generate_config["n"]
+            tools_token_filter = ChatModelMixin._tools_token_filter(self.model_family)
             async for _request_output in results_generator:
                 chunk = self._convert_request_output_to_completion_chunk(
                     request_id=request_id,
                     model=self.model_uid,
                     request_output=_request_output,
                 )
+
                 for i, choice in enumerate(chunk["choices"]):
                     delta = choice["text"][len(previous_texts[i]) :]
                     previous_texts[i] = choice["text"]
                     choice["text"] = delta
+
+                if tools:
+                    # only handle the first choice
+                    choice = chunk["choices"][0]
+                    if choice["finish_reason"] is not None:
+                        choice["text"] = previous_texts[0]
+                        _content, func, args = ChatModelMixin._eval_tool_arguments(
+                            self.model_family, chunk, tools
+                        )
+                        choice["text"] = ""
```

Comment:
Why `text` field becomes empty string when it is the `tool call` results? I think it would be good to keep the origin text information.

Reply:
The main reason for setting the content to empty ('') or None is because I was testing with FastGPT, which initially checks if the content is None. Only if the content is indeed None, does it proceed to determine whether a tool has been invoked.

https://github.com/labring/FastGPT/blob/5cfa43287f16b441f24ae5435e374fdf92c52153/packages/service/core/workflow/dispatch/agent/runTool/toolChoice.ts#L334

Of course their code may not make complete sense.